### PR TITLE
IMPRO-1568: Randomised list of cli inputs.

### PR DESCRIPTION
**Pull request by neilCrosswaite**
Due to extra task on the ticket.
### Commit 1 [287eb4f ](https://github.com/metoppv/improver/commit/287eb4f0238a290e9501bef4f1fe71e532eae79c)

Address the use of a string for a binary value.

### Commit 2 and 3 [5e16414](https://github.com/metoppv/improver/commit/5e164145b55d7b586bb8f02d9856ed60e6a9b911) & [d25824c](https://github.com/metoppv/improver/commit/d25824c9d42c251907f6ff027d6c7afacccc11e5)

Address makeing the individual plugins extract the required cubes they need from a CubeList.
- WetBulbTemperature uses all [CF standard names](http://cfconventions.org/Data/cf-standard-names/70/build/cf-standard-name-table.html)
- PhaseChangeLevel uses `wet_bulb_temperature_integral` which is not a CF standard.

### Commit 4+ [167b07d](https://github.com/metoppv/improver/pull/1146/commits/167b07d0855a6422846396a0d06970d6d71753c6) [a6fb83](https://github.com/metoppv/improver/pull/1146/commits/a6fbf83c402fed1771032a076e41e6554a0db540) [021dc71](https://github.com/metoppv/improver/pull/1146/commits/021dc712b0009a91d1fa471af835512ae588d588)

Use generator instead of a function
Adds randomised order to a test.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- improver/__init__.py  5
         

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver/__init__.py  5
         

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver/__init__.py  5
         

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver/__init__.py  5
         

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- improver/__init__.py  5
         

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

File: improver/__init__.py [hunk 1]
```diff
@@ -53,3 +56,47 @@ def __call__(self, *args, **kwargs):
...
+
+        if isinstance(cubes, list):
+            cubes = CubeList(cubes)
+        if len(cubes) != len(names_to_extract):
+            raise ValueError(
+                f'Expected {len(names_to_extract)} cubes, found {len(cubes)}')
+
+        missing = []
+        extracted = []
+        for i in names_to_extract:
```
Comment (neilCrosswaite) [inline improver/__init__.py]: Change variable

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 2]
```diff
@@ -378,25 +378,32 @@ def create_wet_bulb_temperature_cube(
...
+                        Cube of air pressures.
 
         Returns:
             iris.cube.Cube:
                 Cube of wet bulb temperature (K).
         """
+
+        # pylint: disable=unbalanced-tuple-unpacking
+        temperature, relative_humidity, pressure = self._extract_cubes(
+            cubes, ["air_temperature", "relative_humidity", "air_pressure"])
```
Comment (LaurenceBeard) [inline improver/psychrometric_calculations/psychrometric_calculations.py]: Can we avoid imbedding iris functions, all we are doing is reraising an iris ConstraintMismatchError, which is probably more descriptive than a TypeError?
```
names = ("air_temperature", "relative_humidity", "air_pressure")
tuple(CubeList(f).extract_strict(n) for n in names)
```

Comment (LaurenceBeard) [inline improver/psychrometric_calculations/psychrometric_calculations.py]: Can we avoid imbedding iris functions, all we are doing is reraising an iris ConstraintMismatchError, which is probably more descriptive than a TypeError?
```
temperature, relative_humidity, pressure = tuple(CubeList(f).extract_strict(n) for n in 
                                                                ("air_temperature", "relative_humidity", "air_pressure"))
```

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

File: improver/psychrometric_calculations/psychrometric_calculations.py (hunk 2 above)
Comment (neilCrosswaite) [reply improver/psychrometric_calculations/psychrometric_calculations.py]: Done! :grinning:

File: improver_tests/utilities/test_mathematical_operations.py [hunk 3]
```diff
@@ -105,9 +91,9 @@ def setUp(self):
         self.descending_cube = _set_up_height_cube(
             self.descending_height_points, ascending=False)
         self.plugin_positive = Integration(
-            "height", direction_of_integration="positive")
+            "height", positive_integration=True)
         self.plugin_negative = Integration(
-            "height", direction_of_integration="negative")
+            "height")
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver_tests/utilities/test_mathematical_operations.py]: Thank you. done!

File: improver/utilities/mathematical_operations.py [hunk 4]
```diff
@@ -65,7 +65,7 @@ def __init__(self, coord_name_to_integrate,
                 Point at which to end the integration.
                 Default is None. If end_point is None, integration will
                 continue until the last available point.
-            direction_of_integration (str):
+            positive_integration (bool):
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver/utilities/mathematical_operations.py]: Done, had been done before but i gitt'd it up. oops.

File: improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py [hunk 5]
```diff
@@ -622,12 +641,46 @@ def test_interpolation_from_sea_points(self):
...
-            self.wet_bulb_temperature_cube, self.wet_bulb_integral_cube,
-            orog, land_sea)
+            CubeList([self.wet_bulb_temperature_cube,
+                      self.wet_bulb_integral_cube,
+                      orog, land_sea]))
         expected = np.ones((3, 3, 3), dtype=np.float32) * 65.88566
         self.assertIsInstance(result, iris.cube.Cube)
         self.assertArrayAlmostEqual(result.data, expected)
 
+    def test_too_many_cubes(self):
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py]: Empty lists are good to test differently due to the way python converts an empty list to a bool.

File: improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py [hunk 6]
```diff
@@ -622,12 +641,46 @@ def test_interpolation_from_sea_points(self):
...
+    def test_empty_cube_list(self):
+        """
+        Tests that an error is raised if there is an empty list.
+        """
+        msg = "Expected 4"
+        with self.assertRaisesRegex(ValueError, msg):
+            PhaseChangeLevel(phase_change='snow-sleet').process(
+                CubeList([]))
+
+    def test_wrong_cube(self):
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py]: Sure

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 7]
```diff
@@ -378,25 +379,40 @@ def create_wet_bulb_temperature_cube(
...
+                            "relative_humidity",
+                            "air_pressure"]
+        if len(cubes) != len(names_to_extract):
+            raise ValueError(
+                f'Expected {len(names_to_extract)} cubes, found {len(cubes)}')
+
+        temperature, relative_humidity, pressure = tuple(
+            CubeList(cubes).extract_strict(n) for n in names_to_extract)
+
+        # temperature, relative_humidity, pressure = self._extract_cubes(
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver/psychrometric_calculations/psychrometric_calculations.py]: Done

File: improver/psychrometric_calculations/psychrometric_calculations.py [hunk 8]
```diff
@@ -378,25 +379,40 @@ def create_wet_bulb_temperature_cube(
...
+                    relative_humidity (iris.cube.Cube):
+                        Cube of relative humidities.
+                    pressure (iris.cube.Cube):
+                        Cube of air pressures.
 
         Returns:
             iris.cube.Cube:
                 Cube of wet bulb temperature (K).
         """
+        names_to_extract = ["air_temperature",
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver/psychrometric_calculations/psychrometric_calculations.py]: But that is out of scope for the rest of the class.

File: improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py [hunk 9]
```diff
@@ -622,12 +641,46 @@ def test_interpolation_from_sea_points(self):
...
-            orog, land_sea)
+            CubeList([self.wet_bulb_temperature_cube,
+                      self.wet_bulb_integral_cube,
+                      orog, land_sea]))
         expected = np.ones((3, 3, 3), dtype=np.float32) * 65.88566
         self.assertIsInstance(result, iris.cube.Cube)
         self.assertArrayAlmostEqual(result.data, expected)
 
+    def test_too_many_cubes(self):
+        """
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py]: Also changed the word "arguments" to "cubes".

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

File: improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py [hunk 10]
```diff
@@ -35,6 +35,8 @@
 import iris
 import numpy as np
 from cf_units import Unit
+from iris.cube import CubeList
+from iris.exceptions import ConstraintMismatchError
```
Comment (benfitzpatrick) [inline improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py]: Codacy Issue found: [Unused ConstraintMismatchError imported from iris.exceptions](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853)

File: improver_tests/psychrometric_calculations/test_WetBulbTemperature.py [hunk 11]
```diff
@@ -35,7 +35,8 @@
 import iris
 import numpy as np
 from cf_units import Unit
-from iris.cube import Cube
+from iris.cube import Cube, CubeList
+from iris.exceptions import ConstraintMismatchError
```
Comment (benfitzpatrick) [inline improver_tests/psychrometric_calculations/test_WetBulbTemperature.py]: Codacy Issue found: [Unused ConstraintMismatchError imported from iris.exceptions](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

File: improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py [hunk 12]
```diff
@@ -574,8 +575,24 @@ def test_snow_sleet_phase_change(self):
...
+        self.assertArrayAlmostEqual(result.data, self.expected_snow_sleet)
+
+    def test_snow_sleet_phase_change_reorder_cubes(self):
+        """Same test as test_snow_sleet_phase_change but the cubes are in a
+        different order"""
+        self.orog.data[1, 1] = 100.0
+        result = PhaseChangeLevel(phase_change='snow-sleet').process(
+            CubeList([self.wet_bulb_integral_cube,
+                      self.wet_bulb_temperature_cube,
+                      self.orog, self.land_sea])
```
Comment (bayliffe) [inline improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py]: These could be more mis-ordered 😄

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Comment by benfitzpatrick** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Clones added
============
- improver_tests/psychrometric_calculations/test_PhaseChangeLevel.py  2
         

Clones removed
==============
+ improver_tests/utilities/test_mathematical_operations.py  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=4948853&bid=16356800)

**Pull request merged**